samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -0,0 +1,288 @@
+}
+
+# <NOT ...>
+ZIL_BUILTINS[:NOT] = define_for_evaled_arguments { |arguments|
+  raise FunctionError, "NOT only supported with 1 argument!" if arguments.length != 1
+  arguments[0] == false
+}
+
+# <AND ...> (FSUBR)
+ZIL_BUILTINS[:AND] = lambda { |arguments, context|
+  result = false
+  arguments.each { |a|
+    result = eval_zil(a, context)
+    return false if result == false
+  }
+  result
+}
+
+#! shouldn't those just be doable with `Enumerable#all?`
+# <AND? ...> (SUBR)
+ZIL_BUILTINS[:AND?] = define_for_evaled_arguments { |arguments|
+  result = false
+  arguments.each { |a|
+    result = a
+    return false if a == false
+  }
+  result
+}
+
+# or with `Enumerable#any?` in this case
+# <OR ...>  (FSUBR)
+ZIL_BUILTINS[:OR] = lambda { |arguments, context|
+  arguments.each { |a|
+    result = eval_zil(a, context)
+    return result unless result == false
+  }
+  false
+}
+
+# <OR? ...>  (SUBR)
+ZIL_BUILTINS[:OR?] = define_for_evaled_arguments { |arguments|
+  arguments.each { |a|
+    return a unless a == false
+  }
+  false
+}
+
+# <COND () ...> (FSUBR)
+ZIL_BUILTINS[:COND] = lambda { |arguments, context|
+  # COND goes walking down its clauses, EVALing the first element of each clause,
+  # looking for a non-FALSE result. As soon as it finds a non-FALSE, it forgets
+  # about all the other clauses and evaluates, in order, the other elements of
+  # the current clause and returns the last thing it evaluates.
+  #
+  raise FunctionError, "COND requires at least one parameter!" unless arguments.length > 0
+
+  i = 0
+  while i < arguments.length
+    clause = arguments[i]
+
+    raise FunctionError, "Arguments to COND must be of List type!" unless clause.class == Syntax::List
+    raise FunctionError, "COND clauses require at least 1 item!" unless clause.elements.length > 0
+
+    clause_eval = eval_zil(clause.elements[0], context)
+    if clause_eval != false # eval 0th element for non-false
+      clause.elements.drop(1).each { |clause_element|
+        clause_eval = eval_zil(clause_element, context)
+      }
+      return clause_eval
+    end
+    i += 1
+  end
+
+  false
+}
+
+ZIL_BUILTINS[:OBJECT] = lambda { |arguments, context|
+  # Objects have a name, and a list of properties and values
+
+  raise FunctionError, "OBJECT requires at least a name!" unless arguments.length > 0
+
+  # not sure if this is needed
+  raise FunctionError, "OBJECT requires properties" unless arguments.length > 1
+
+  object_name, *object_properties = arguments
+
+  object = { name: object_name, properties: {} }
+
+  object_properties.each do |property|
+    raise FunctionError, "Parameters to object must be in list form" unless property.class == Syntax::List
+    raise FunctionError, "OBJECT properties require a name and values!" unless property.elements.length > 1
+
+    property_name, *property_values = property.elements
+
+    object[:properties][property_name] = property_values.length == 1 ? property_values[0] : property_values
+  end
+
+  object
```

Comment:
The object needs to be saved/defined somewhere, if we just return it the ZIL interpreter it's just going to garbage collect it.